onnxruntime/core/platform/windows/env.cc

```diff
@@ -148,28 +171,141 @@ class WindowsEnv : public Env {
     return processorCoreCount;
   }
 
-  std::vector<size_t> GetThreadAffinityMasks() const override {
-    auto generate_vector_of_n = [](int n) {
-      std::vector<size_t> ret(n);
-      std::iota(ret.begin(), ret.end(), 0);
-      return ret;
-    };
-    // Indeed 64 should be enough. However, it's harmless to have a little more.
-    SYSTEM_LOGICAL_PROCESSOR_INFORMATION buffer[256];
-    DWORD returnLength = sizeof(buffer);
-    if (GetLogicalProcessorInformation(buffer, &returnLength) == FALSE) {
-      return generate_vector_of_n(std::thread::hardware_concurrency());
+  size_t GetDefaultThreadpoolSetting(std::vector<size_t>& affinities) const override {
+    std::vector<KAFFINITY> kAffinities;
+    KAFFINITY group_id = 0;
+    for (const auto& group_info : this->group_info_vec_) {
+      size_t bit_offset = 0;
+      constexpr size_t bit_limit = sizeof(KAFFINITY) << 3;
+      KAFFINITY active_processor_mask = group_info.ActiveProcessorMask;
+      for (int i = 0; i < (group_info.ActiveProcessorCount >> 1); ++i) {
+        KAFFINITY processor_mask = 0;
+        size_t processor_count = 0;
+        while (processor_count < 2 && bit_offset < bit_limit) {
+          auto current_bit = (0x1 << bit_offset);
+          if (active_processor_mask & current_bit) {
+            processor_mask |= current_bit;
+            processor_count++;
+          }
+          bit_offset++;
+        }
+        kAffinities.push_back(group_id);
+        kAffinities.push_back(processor_mask);
+      }
+      group_id++;
     }
-    std::vector<size_t> ret;
-    int count = (int)(returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION));
-    for (int i = 0; i != count; ++i) {
-      if (buffer[i].Relationship == RelationProcessorCore) {
-        ret.push_back(buffer[i].ProcessorMask);
+    if (kAffinities.empty()) {
+      return std::thread::hardware_concurrency() >> 1;
+    }
+    size_t num_of_size_t = kAffinities.size() * sizeof(KAFFINITY) / sizeof(size_t);
+    affinities.clear();
+    affinities.resize(num_of_size_t);
+    memcpy(affinities.data(), kAffinities.data(), num_of_size_t * sizeof(size_t));
+    return kAffinities.size();
+  }
+
+  std::pair<KAFFINITY, KAFFINITY> GetGroupAffinity(int procssor_from, int processor_to) {
```

Comment:
At what point do we subtract 64 for processor id > 63?

Reply:
We do not do subtraction for the purpose. Instead we loop through every group and and account the processors found in each group, when the counter matched one of the processor id, a mask will be set.